### PR TITLE
Fix bounding box leaf collision check

### PR DESCRIPTION
**Pull request by jorgensd**
Reported by @nate-sime 
The following MWE produces the wrong collision for the global_bb_tree
```python
import numpy as np
import dolfinx
from mpi4py import MPI

mesh = dolfinx.mesh.create_unit_square(MPI.COMM_WORLD, 10, 10)

xp = np.array([[100.0, 100.0, 0.0]], dtype=np.double)

tree = dolfinx.geometry.bb_tree(mesh, mesh.topology.dim, padding=1e-4)
global_tree = tree.create_global_tree(mesh.comm)


def pprint(*msg):
    print(f"[{mesh.comm.rank}]: " + " ".join(map(str, msg)), flush=True)


pprint(dolfinx.common.git_commit_hash)
pprint(dolfinx.geometry.compute_collisions_points(tree, xp))
pprint(dolfinx.geometry.compute_collisions_points(global_tree, xp))

```
giving the following in serial:
```bash
[0]: f488f4ecc687d0065315bc9a74939634b18e11ff
[0]: <AdjacencyList> with 1 nodes
  0: []

[0]: <AdjacencyList> with 1 nodes
  0: [0 ]
```
and the correct result in parallel
```bash
[0]: f488f4ecc687d0065315bc9a74939634b18e11ff
[0]: <AdjacencyList> with 1 nodes
  0: []

[0]: <AdjacencyList> with 1 nodes
  0: []

[1]: f488f4ecc687d0065315bc9a74939634b18e11ff
[1]: <AdjacencyList> with 1 nodes
  0: []

[1]: <AdjacencyList> with 1 nodes
  0: []
```

This is due to a bug in `_compute_collisions_point` where we don't check if the leaf node actually collides with the point.
@massimiliano-leoni could you see if this impacts your mesh interpolation (in serial)? As the `global_bb_tree` is used it would add any interpolation point from the other mesh to the process, even if it doesn't collide.

### CODE REVIEW

File: cpp/dolfinx/geometry/utils.h [hunk 1]
```diff
@@ -303,7 +303,7 @@ void _compute_collisions_point(const geometry::BoundingBoxTree<T>& tree,
     const std::array<int, 2> bbox = tree.bbox(next);
     next = -1;
 
-    if (is_leaf(bbox))
+    if ((is_leaf(bbox)) && point_in_bbox(tree.get_bbox(next), p))
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:306]: We and `and` rather than `&&`.

File: cpp/dolfinx/geometry/utils.h [hunk 2]
```diff
@@ -683,7 +683,16 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
   // Compute collisions:
   // For each point in `x` get the processes it should be sent to
   graph::AdjacencyList collisions = compute_collisions(global_bbtree, points);
-
+  for (std::size_t i = 0; i < collisions.num_nodes(); ++i)
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:686]: Remove debug output.

File: cpp/dolfinx/geometry/utils.h [hunk 3]
```diff
@@ -711,7 +720,9 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
   for (std::size_t i = 0; i < points.size() / 3; ++i)
     for (auto p : collisions.links(i))
       send_sizes[rank_to_neighbor[p]] += 3;
-
+  for (auto rank : send_sizes)
```
Comment (garth-wells) [inline cpp/dolfinx/geometry/utils.h:723]: Remove.

**Review by garth-wells — CHANGES_REQUESTED** [review]
Could the PR have a more descriptive title?

And could a test be added?

File: cpp/dolfinx/geometry/utils.h [hunk 4]
```diff
@@ -679,17 +679,19 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
   BoundingBoxTree bb(mesh, tdim, cells, padding);
   BoundingBoxTree midpoint_tree = create_midpoint_tree(mesh, tdim, cells);
   BoundingBoxTree global_bbtree = bb.create_global_tree(comm);
+  std::cout << "HERE!\n";
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.h:682]:
```suggestion
```

File: cpp/dolfinx/geometry/utils.h [hunk 5]
```diff
@@ -679,17 +679,19 @@ determine_point_ownership(const mesh::Mesh<T>& mesh, std::span<const T> points)
...
-
   // Get unique list of outgoing ranks
   std::vector<std::int32_t> out_ranks = collisions.array();
   std::sort(out_ranks.begin(), out_ranks.end());
   out_ranks.erase(std::unique(out_ranks.begin(), out_ranks.end()),
                   out_ranks.end());
-
+  for (auto rank : out_ranks)
+    std::cout << rank << " ";
+  std::cout << "\n";
```
Comment (jorgensd) [inline cpp/dolfinx/geometry/utils.h:694]:
```suggestion

```

**Review by garth-wells — APPROVED** [review]

**Review by IgorBaratta — APPROVED** [review]

**Review by nate-sime — APPROVED** [review]

**Comment by jorgensd** [comment]
@garth-wells  Test added + minor simplification. Feel free to merge this once the CI passes.

**Pull request merged**